baseplate/diagnostics/tracing.py

```diff
@@ -0,0 +1,232 @@
+                                           self.recorder)
+        server_span.register(observer)
+
+
+class TraceSpanObserver(SpanObserver):
+    """Span recording observer for outgoing request child spans.
+
+    This observer implements the client-side span portion of a
+    Zipkin request trace.
+    """
+    def __init__(self, service_name, span, recorder):
+        self.service_name = service_name
+        self.recorder = recorder
+        self.span = span
+        self.start = None
+        self.end = None
+        self.elapsed = None
+        self.binary_annotations = []
+        super(TraceSpanObserver, self).__init__()
+
+    def on_start(self):
+        self.start = current_epoch_microseconds()
+        self.client_send = self.start
+
+    def on_finish(self, exc_info):
+        self.end = current_epoch_microseconds()
+        self.elapsed = self.end - self.start
+        self.record()
+
+    def _create_time_annotation(self, annotation_type, timestamp):
+        """Create Zipkin-compatible Annotation for a span.
+
+        This should be used for generating span annotations with a time component,
+        e.g. the core "cs", "sr", "ss", and "sr" Zipkin Annotations
+        """
+        endpoint_info = {
+            'serviceName': self.service_name,
+            'ipv4': socket.gethostbyname(socket.gethostname()),
+        }
+        return {
+            'endpoint': endpoint_info,
+            'timestamp': timestamp,
+            'value': annotation_type,
+        }
+
+    def _create_binary_annotation(self, ):
+        """Create Zipkin-compatible BinaryAnnotation for a span.
+
+        This should be used for generating span annotations that
+        do not have a time component, e.g. URI, arbitrary request tags
+        """
+        # TODO: Implement.
+        pass
+
+    def _to_span_obj(self, annotations, binary_annotations):
+        span = {
+            "traceId": self.span.trace_id,
+            "name": self.span.name,
+            "id": self.span.id,
+            "timestamp": self.start,
+            "duration": self.elapsed,
+            "annotations": annotations,
+            "binary_annotations": binary_annotations,
+        }
+
+        span['parentId'] = self.span.parent_id or 0
+        return span
+
+    def _serialize(self):
+        """Serialize span information into Zipkin-accepted format."""
+        annotations = []
+
+        annotations.append(
+            self._create_time_annotation(
+                ANNOTATIONS['CLIENT_RECEIVE'],
+                self.start
+            )
+        )
+
+        annotations.append(
+            self._create_time_annotation(
+                ANNOTATIONS['CLIENT_SEND'],
+                self.end
+            )
+        )
+
+        return self._to_span_obj(annotations, self.binary_annotations)
+
+    def record(self):
+        """Record serialized span."""
+        serialized = self._serialize()
+        t = threading.Thread(target=self.recorder.send,
```

Comment:
Having an intermediary queue + worker adds an extra level of complexity to manage. I'm working under the assumption that greenlets will be the default here so that this will remain relatively inexpensive under high request load, and that this is a simple enough operation that greenlets will not be long-lived. 

That being said, it might be safer from a resource perspective to use a queue + worker as system resources will remain constant under higher load, aside from the queue memory impact due to back up. 

There are a couple of zipkin implementations on both sides of this but after thinking a bit more I'm going to err on the side of safety for this first pass.